lib/ts/recipe/totp/types.ts

```diff
@@ -0,0 +1,143 @@
+    };
+};
+
+export type RecipeInterface = {
+    createDevice: (input: {
+        userId: string;
+        deviceName: string;
+        skew?: number;
+        period?: number;
+        userContext: any;
+    }) => Promise<{ status: "OK"; qr: string } | { status: "DEVICE_ALREADY_EXISTS_ERROR" }>;
+    verifyCode: (input: {
+        userId: string;
+        totp: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK" | "INVALID_TOTP_ERROR" | "TOTP_NOT_ENABLED_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    verifyDevice: (input: {
+        userId: string;
+        deviceName: string;
+        totp: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK" | "INVALID_TOTP_ERROR" | "TOTP_NOT_ENABLED_ERROR" | "UNKNOWN_DEVICE_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    updateDevice: (input: {
+        userId: string;
+        existingDeviceName: string;
+        newDeviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK" | "TOTP_NOT_ENABLED_ERROR" | "DEVICE_ALREADY_EXISTS_ERROR" | "UNKNOWN_DEVICE_ERROR" }>;
+    removeDevice: (input: {
+        userId: string;
+        deviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK" | "TOTP_NOT_ENABLED_ERROR" }>;
+    listDevices: (input: {
+        userId: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; devices: { name: string; period: number; skew: number; verified: boolean }[] }
+        | { status: "TOTP_NOT_ENABLED_ERROR" }
+    >;
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    appInfo: NormalisedAppinfo;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    createDevicePOST?: (input: {
+        session: SessionContainer;
+        deviceName: string;
+        period: number;
+        skew: number;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<{ status: "OK"; qr: string } | { status: "DEVICE_ALREADY_EXISTS_ERROR" }>; // TODO: Consider seperately mentioning templates vars used in qr (issuerName, secret, provisioner, etc)
+    verifyCodePOST?: (input: {
+        session: SessionContainer;
+        totp: string;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK" | "INVALID_TOTP_ERROR" | "TOTP_NOT_ENABLED_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    verifyDevicePOST?: (input: {
+        session: SessionContainer;
+        deviceName: string;
+        totp: string;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK" | "INVALID_TOTP_ERROR" | "TOTP_NOT_ENABLED_ERROR" | "UNKNOWN_DEVICE_ERROR" }
```

Comment:
This API should not be called at all in case TOTP_NOT_ENABLED_ERROR. Therefore, if the rceipe function returns this, we can throw a JS error from this API - so no need to return this status code?